src/components/views/rooms/RoomKnocksBar.tsx

```diff
@@ -0,0 +1,152 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { EventTimeline, JoinRule, MatrixError, Room, RoomStateEvent } from "matrix-js-sdk/src/matrix";
+import React, { ReactElement, ReactNode, useCallback, useState, VFC } from "react";
+
+import { Icon as CheckIcon } from "../../../../res/img/feather-customised/check.svg";
+import { Icon as XIcon } from "../../../../res/img/feather-customised/x.svg";
+import dis from "../../../dispatcher/dispatcher";
+import { useTypedEventEmitterState } from "../../../hooks/useEventEmitter";
+import { _t } from "../../../languageHandler";
+import Modal from "../../../Modal";
+import MemberAvatar from "../avatars/MemberAvatar";
+import ErrorDialog from "../dialogs/ErrorDialog";
+import { RoomSettingsTab } from "../dialogs/RoomSettingsDialog";
+import AccessibleButton from "../elements/AccessibleButton";
+import Heading from "../typography/Heading";
+
+export const RoomKnocksBar: VFC<{ room: Room }> = ({ room }) => {
+    const client = room.client;
+    const userId = client.getUserId() || "";
+    const canInvite = room.canInvite(userId);
+    const member = room.getMember(userId);
+    const state = room.getLiveTimeline().getState(EventTimeline.FORWARDS);
+    const canKick = member && state ? state.hasSufficientPowerLevelFor("kick", member.powerLevel) : false;
+
+    const handleApprove = (userId: string): void => {
+        setDisabled(true);
+        client.invite(room.roomId, userId).catch(onError);
+    };
+
+    const handleDeny = (userId: string): void => {
+        setDisabled(true);
+        client.kick(room.roomId, userId).catch(onError);
+    };
+
+    const handleOpenRoomSettings = (): void =>
+        dis.dispatch({ action: "open_room_settings", room_id: room.roomId, initial_tab_id: RoomSettingsTab.People });
+
+    const onError = (error: MatrixError): void => {
+        setDisabled(false);
+        Modal.createDialog(ErrorDialog, { title: error.name, description: error.message });
+    };
+
+    const [disabled, setDisabled] = useState(false);
+    const knockMembers = useTypedEventEmitterState(
+        room,
+        RoomStateEvent.Members,
+        useCallback(() => room.getMembersWithMembership("knock"), [room]),
+    );
+    const knockMembersCount = knockMembers.length;
+
+    if (room.getJoinRule() !== JoinRule.Knock || knockMembersCount === 0 || (!canInvite && !canKick)) return null;
+
+    let buttons: ReactElement = (
+        <AccessibleButton
+            className="mx_RoomKnocksBar_action"
+            kind="primary"
+            onClick={handleOpenRoomSettings}
+            title={_t("action|view")}
+        >
+            {_t("action|view")}
+        </AccessibleButton>
+    );
+    let names: string = knockMembers
+        .slice(0, 2)
+        .map((knockMember) => knockMember.name)
+        .join(", ");
```

Comment:
Is this how we make lists of names in other parts of the app? It strikes me as potentially problematic for some languages, but if it's a pattern we use elsewhere then it's fine to repeat here.

Reply:
It is similar to
https://github.com/matrix-org/matrix-react-sdk/blob/6aa86a858f68a29bbda9438dc4a41a3915345d3c/src/WhoIsTyping.ts#L51
https://github.com/matrix-org/matrix-react-sdk/blob/6aa86a858f68a29bbda9438dc4a41a3915345d3c/src/WhoIsTyping.ts#L72

Reply:
Cool, let's leave it then. I imagine at some point we may have to revisit, but that's not on you.